### PR TITLE
Add support for Server Variables and relevant tests in Chai OpenAPI Response Validator

### PR DESCRIPTION
**Pull request by JonnySpruce**
I believe this PR resolves https://github.com/RuntimeTools/OpenAPIValidators/issues/89, by adding support for server variables. I have added all tests and code for the Chai plugin - the Jest plugin is still TODO at this stage, although now the code is written it should be very easy to port across.

@Nitwel - can you confirm this resolves the issue for you?

### CODE REVIEW

**Comment by rwalle61** [comment]
Cheers, let's get the Chai code ready to merge, then port it to Jest

File: packages/chai-openapi-response-validator/lib/openapi-validator/lib/classes/OpenApi3Spec.js [hunk 1]
```diff
@@ -108,4 +112,43 @@ class OpenApi3Spec extends AbstractOpenApiSpec {
...
+  const serverVars = Object.keys(server.variables);
+
+  serverVars.forEach(serverVar => {
+    serverVarMap.set(serverVar, server.variables[serverVar].enum);
+  });
+
+  return serverVarMap;
+}
+
+function replaceServerVarsInURLs (serverURLs, serverVarMap) {
```
Comment (rwalle61) [inline packages/chai-openapi-response-validator/lib/openapi-validator/lib/classes/OpenApi3Spec.js:129]: Although the tests prove this function works, ideally I'd prefer a simpler, more intuitive function if possible (esp without recursion). 

I'll re-review this function after we've resolved the other comments, to see if we can think of a way to write it so that a newcomer (or I) can understand it easily :)

**Review by rwalle61 — COMMENTED** [review]
Thanks very much for this, great to see plenty of test cases. I think we can ignore server variables in the host, and we may need to add one more test case to cover an edge case, but otherwise the suggestions are minor!

File: packages/chai-openapi-response-validator/test/assertions/satisfyApiSpec/specsDefineServersDifferently.test.js [hunk 2]
```diff
@@ -493,4 +493,127 @@ describe('Using OpenAPI 3 specs that define servers differently', () => {
...
+      );
+      chai.use(chaiResponseValidator(pathToApiSpec));
+    });
+
+    describe("res.req.path matches the server path with a single variable", () => {
+      const res = {
+        status: 200,
+        req: {
+          method: 'GET',
+          path: 'https://test.com:1234/test/responseBody/string',
```
[the remark this answers is not in the document]
Comment (JonnySpruce) [reply packages/chai-openapi-response-validator/test/assertions/satisfyApiSpec/specsDefineServersDifferently.test.js:511]: You're absolutely right, about to push changes to the tests which reflect real request structures. I have left the server variables in the spec to ensure that we work as expected when they are included. Agree we only need to change the variables after the paths - is detecting the start of the path as simple as looking for the first single `/` or are there some edge cases we would also need to be able to detect?

File: packages/chai-openapi-response-validator/test/assertions/satisfyApiSpec/specsDefineServersDifferently.test.js [hunk 3]
```diff
@@ -493,4 +493,127 @@ describe('Using OpenAPI 3 specs that define servers differently', () => {
...
+        },
+        body: 'valid body (string)',
+      };
+
+      it('fails', () => {
+        const assertion = () => expect(res).to.satisfyApiSpec;
+        expect(assertion).to.throw(
+          AssertionError,
+          `'/nonExistentEndpointPath' matches servers ${inspect([
+            'https://test.com:1234',
```
[the remark this answers is not in the document]
Comment (JonnySpruce) [reply packages/chai-openapi-response-validator/test/assertions/satisfyApiSpec/specsDefineServersDifferently.test.js:581]: I think we might actually want to be a little more specific than this, and it goes hand in hand with the comment you made elsewhere about only changing the server variables in the path.

I.e. in my tests I have a server with a port variable, and a path variable: `https://test.com:{portVariable}/{pathVariable}`. When I make a request with the path `/defaultPathVariable` at the moment I would get back:
```
      '/defaultPathVariable' matches servers [
        'https://test.com:1234/defaultPathVariable',
        'https://test.com:5678/defaultPathVariable'
      ]
```
Obviously the multiple servers with different port numbers is excessive, however if we changed it to 
```
      '/defaultPathVariable' matches servers [
        'https://test.com:{portVariable}/{pathVariable}'
      ]
```
Then it's actually too vague, since in reality the request only matches one of the `pathVariable` enums: `defaultPathVariable`, not all `pathVariable` enums.

Instead, if we do the work to only change the server variables in the path, then it should be possible to give the following error message, which I think is the preferred solution:

 ```
      '/defaultPathVariable' matches servers [
        'https://test.com:{portVariable}/defaultPathVariable'
      ]
```

File: packages/chai-openapi-response-validator/lib/openapi-validator/lib/classes/OpenApi3Spec.js (hunk 1 above)
Comment (JonnySpruce) [reply packages/chai-openapi-response-validator/lib/openapi-validator/lib/classes/OpenApi3Spec.js]: Happy to work on this another way, but not exactly sure how else to approach it, if you have any suggestions I'd be happy to take another crack at it.

**Comment by JonnySpruce** [comment]
I've pushed another commit which resolves the minor comments you had with the code. Happy to also do the work for splitting the host and path parts of the server URL, but want to make sure I'm doing it correctly and not somehow breaking other things by splitting in the wrong place in certain scenarios.

File: packages/chai-openapi-response-validator/test/assertions/satisfyApiSpec/specsDefineServersDifferently.test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (rwalle61) [reply packages/chai-openapi-response-validator/test/assertions/satisfyApiSpec/specsDefineServersDifferently.test.js]: you may be able to make use of `url.parse` e.g. https://github.com/RuntimeTools/OpenAPIValidators/blob/48eddb49bdc5a18af4de01bfb1055d315335ff26/packages/chai-openapi-response-validator/lib/openapi-validator/lib/utils.js#L11

File: packages/chai-openapi-response-validator/test/assertions/satisfyApiSpec/specsDefineServersDifferently.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (rwalle61) [reply packages/chai-openapi-response-validator/test/assertions/satisfyApiSpec/specsDefineServersDifferently.test.js]: well put, I agree!

**Comment by rwalle61** [comment]
More thoughts, having looked at this more closely:

Given:
1. our updated understanding that we don't need to worry about server variables before the path when finding a match for `req.path`,
2. [the `default` value is required](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#fixed-fields-5), so in our test openapi.yml we need server variables without `enum`.
3. also, [only the default value is required](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#fixed-fields-5), so we can simplify the server variables in our test openapi.yml
4. ["If the enum is defined, the [default] value SHOULD exist in the enum's values."](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#fixed-fields-5). The "SHOULD" means that although they recommend that the default value is one of the enum values, it doesn't have to be. I think this will be easy to implement.

We could rename our test cases roughly like this, to clarify exactly what we are testing (and add any that are missing from this PR):

```
'res.req.path matches a server with server variables only before the path'

'res.req.path matches a server with a server variable in the path (matches the default value)'

'res.req.path matches a server with a server variable in the path (matches a non-default value)'

'res.req.path matches a server with multiple server variables in the path'

'res.req.path matches a server with a server variable in the path, but matches no endpoint paths'

"res.req.path does not match the default server base path ('/') nor any servers"
```

Does our test openapi.yml need to be any more complex than this?
```
openapi: 3.0.0
info:
  title: Example OpenApi 3 spec
  description: Defines servers with server variables
  version: 0.1.0
servers:
  - url: https://{hostVariable1}-{hostVariable2}.com:{portVariable}/
    description: contains only server variables before the path
    variables:
      hostVariable1:
        default: ignoreMe
      hostVariable2:
        default: ignoreMeAlso
      portVariable:
        default: '1234'
  - url: /{variableInPath}
    description: server url with 1 server variable in the path
    variables:
      variableInPath:
        default: defaultValueForVariableInPath
        enum:
          - enumValue1ForVariableInPath
          - enumValue2ForVariableInPath
  - url: /{firstVariableInPath}/{secondVariableInPath}
    description: server url with multiple server variables in the path
    variables:
      firstVariableInPath:
        default: defaultValueForFirstVariableInPath
      secondVariableInPath:
        default: defaultValueForSecondVariableInPath
paths:
  /endpointPath:
    get:
      responses:
        200:
          description: Response body should be a string
          content:
            application/json:
              schema:
                type: string
```

What do you think?

**Comment by rwalle61** [comment]
Superceded by https://github.com/openapi-library/OpenAPIValidators/pull/222. Thanks @JonnySpruce for your work on this, you'll be credited in the release!
